File: day-2-pm/sandbox/tests/parser_test.cpp
#include "parser.h"

#include <gtest/gtest.h>

using calc::BinOp;
using calc::NodeKind;
using calc::parse;

TEST(ParserTest, SingleNumber) {
    auto ast = parse("7");
    ASSERT_EQ(ast->kind, NodeKind::Number);
    EXPECT_EQ(ast->value, 7);
}

TEST(ParserTest, FlatAdditionIsLeftAssociative) {
    // 1 + 2 + 3 -> ((1 + 2) + 3)
    auto ast = parse("1 + 2 + 3");
    ASSERT_EQ(ast->kind, NodeKind::BinaryOp);
    EXPECT_EQ(ast->op, BinOp::Add);
    ASSERT_EQ(ast->left->kind, NodeKind::BinaryOp);
    EXPECT_EQ(ast->left->op, BinOp::Add);
    EXPECT_EQ(ast->left->left->value, 1);
    EXPECT_EQ(ast->left->right->value, 2);
    ASSERT_EQ(ast->right->kind, NodeKind::Number);
    EXPECT_EQ(ast->right->value, 3);
}

TEST(ParserTest, MulBindsTighterThanAdd) {
    // 1 + 2 * 3 -> (1 + (2 * 3))
    auto ast = parse("1 + 2 * 3");
    ASSERT_EQ(ast->kind, NodeKind::BinaryOp);
    EXPECT_EQ(ast->op, BinOp::Add);
    ASSERT_EQ(ast->left->kind, NodeKind::Number);
    EXPECT_EQ(ast->left->value, 1);
    ASSERT_EQ(ast->right->kind, NodeKind::BinaryOp);
    EXPECT_EQ(ast->right->op, BinOp::Mul);
}

TEST(ParserTest, TrailingTokenThrows) {
    EXPECT_THROW(parse("1 2"), std::runtime_error);
}

TEST(ParserTest, MissingOperandThrows) {
    EXPECT_THROW(parse("1 +"), std::runtime_error);
}

File: day-2-pm/sandbox-java/src/test/java/calc/ParserTest.java
package calc;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.Test;

class ParserTest {

    @Test
    void SingleNumber() {
        Node ast = Parser.parse("7");
        assertEquals(NodeKind.Number, ast.kind);
        assertEquals(7, ast.value);
    }

    @Test
    void FlatAdditionIsLeftAssociative() {
        // 1 + 2 + 3 -> ((1 + 2) + 3)
        Node ast = Parser.parse("1 + 2 + 3");
        assertEquals(NodeKind.BinaryOp, ast.kind);
        assertEquals(BinOp.Add, ast.op);
        assertEquals(NodeKind.BinaryOp, ast.left.kind);
        assertEquals(BinOp.Add, ast.left.op);
        assertEquals(1, ast.left.left.value);
        assertEquals(2, ast.left.right.value);
        assertEquals(NodeKind.Number, ast.right.kind);
        assertEquals(3, ast.right.value);
    }

    @Test
    void MulBindsTighterThanAdd() {
        // 1 + 2 * 3 -> (1 + (2 * 3))
        Node ast = Parser.parse("1 + 2 * 3");
        assertEquals(NodeKind.BinaryOp, ast.kind);
        assertEquals(BinOp.Add, ast.op);
        assertEquals(NodeKind.Number, ast.left.kind);
        assertEquals(1, ast.left.value);
        assertEquals(NodeKind.BinaryOp, ast.right.kind);
        assertEquals(BinOp.Mul, ast.right.op);
    }

    @Test
    void TrailingTokenThrows() {
        assertThrows(CalcException.class, () -> Parser.parse("1 2"));
    }

    @Test
    void MissingOperandThrows() {
        assertThrows(CalcException.class, () -> Parser.parse("1 +"));
    }
}
